File: scorer.py
import enum

import card

CARDS_TO_USE = 5


class HandType(enum.IntEnum):  # Must be ordered by priority descending
    ROYAL_FLUSH = 10
    STRAIGHT_FLUSH = 9
    FOUR_KIND = 8
    FULL_HOUSE = 7
    FLUSH = 6
    STRAIGHT = 5
    THREE_KIND = 4
    TWO_PAIR = 3
    PAIR = 2
    HIGH_CARD = 1


HAND_NAMES = {
    HandType.ROYAL_FLUSH: "royal_flush ",
    HandType.STRAIGHT_FLUSH: "straight_flush",
    HandType.FOUR_KIND: "four_kind",
    HandType.FULL_HOUSE: "full_house",
    HandType.FLUSH: "flush",
    HandType.STRAIGHT: "straight",
    HandType.THREE_KIND: "three_kind",
    HandType.TWO_PAIR: "two_pair",
    HandType.PAIR: "pair",
    HandType.HIGH_CARD: "high_card",
}


def tiebreaker_lt(cards_a, a_ace_hi, cards_b, b_ace_hi):
    cards = [cards_a, cards_b]
    ace_hi = [a_ace_hi, b_ace_hi]
    rankings = [[c.val for c in cs] for cs in cards]
    for i in range(len(rankings)):
        if not ace_hi[i]:
            rankings[i] = [1 if r == 14 else r for r in rankings[i]]
        rankings[i].sort(reverse=True)

    for i in range(min([len(r) for r in rankings])):
        if rankings[0][i] < rankings[1][i]:
            return True
        elif rankings[0][i] > rankings[1][i]:
            return False

    if len(rankings[0]) < len(rankings[1]):
        return True
    elif len(rankings[0]) > len(rankings[1]):
        return False

    return None


class Score:
    def __init__(self, hand_type, hand_cards, cards):
        self.hand_type = hand_type
        self.hand_cards = hand_cards
        self.cards = cards
        self.cards.sort()
        self.non_hand_cards = [
            c for c in self.cards if c not in self.hand_cards
        ]

    def __str__(self):
        return "{ hand=" + HAND_NAMES[self.
                                      hand_type] + " hand_cards=" + " ".join([
                                          str(c) for c in self.hand_cards
                                      ]) + " cards=" + " ".join(
                                          [str(c) for c in self.cards]) + " }"

    def __eq__(self, other):
        return not self < other and not self > other

    def __lt__(self, other):
        if self.hand_type != other.hand_type:
            return self.hand_type < other.hand_type

        # Tiebreaker
        self_ace_hi = True
        other_ace_hi = True
        # Ace is low iff we have a straight with the ace at the bottom (no
        # king in the straight)
        hand_vals = [c.val for c in self.hand_cards]
        if self.hand_type == HandType.STRAIGHT and 13 not in hand_vals:
            self_ace_hi = False
        if other.hand_type == HandType.STRAIGHT and 13 not in hand_vals:
            other_ace_hi = False

        hand_tiebreaker = tiebreaker_lt(self.hand_cards, self_ace_hi,
                                        other.hand_cards, other_ace_hi)
        if hand_tiebreaker is not None:
            return hand_tiebreaker

        hand_size = len(self.hand_cards)
        if hand_size == 5:
            return False

        return not not tiebreaker_lt(self.non_hand_cards[-(5 - hand_size):],
                                     False, other.cards, False)


def find_n_kind(cards, n):
    num_sequential = 1
    for i in reversed(range(len(cards) - 1)):
        if cards[i].val == cards[i + 1].val:
            num_sequential += 1
        else:
            num_sequential = 1
        if num_sequential == n:
            return cards[i:i + n]
    return None


# Returns the cards making up the hand, or None of the hand does not exist
# CARDS MUST BE SORTED
def find_hand(cards, hand):
    assert len(cards) <= 1 or (cards[0].val <= cards[1].val
                               and cards[-2].val <= cards[-1].val
                               ), "Cards passed to find_hand must be sorted"

    if hand == HandType.ROYAL_FLUSH:
        # We have a royal flush iff we have a straight flush with only cards
        # >= 10
        return find_hand([c for c in cards if c.val >= 10],
                         HandType.STRAIGHT_FLUSH)
    elif hand == HandType.STRAIGHT_FLUSH:
        # We do this whole best_straight thing to account for cases where we
        # have 10 cards or more, since in those cases we could have multiple
        # suits with a straight flush and we should pick the best one
        best_straight = None
        for suit in range(card.NUM_SUITS):
            straight = find_hand([c for c in cards if c.suit == suit],
                                 HandType.STRAIGHT)
            if straight and (best_straight is None
                             or straight[-1].val > best_straight[-1].val):
                best_straight = straight
        return best_straight
    elif hand == HandType.FOUR_KIND:
        return find_n_kind(cards, 4)
    elif hand == HandType.FULL_HOUSE:
        three_kind = find_hand(cards, HandType.THREE_KIND)
        if three_kind is not None:
            pair = find_hand([c for c in cards if c not in three_kind],
                             HandType.PAIR)
            if pair is not None:
                return pair + three_kind
        return None
    elif hand == HandType.FLUSH:
        best_flush = None
        for suit in range(card.NUM_SUITS):
            suit_cards = [c for c in cards if c.suit == suit]
            if len(suit_cards) >= 5 and (
                    best_flush is None
                    or suit_cards[-1].val > best_flush[-1].val):
                best_flush = suit_cards[len(suit_cards) - 5:]
    elif hand == HandType.STRAIGHT:
        if len(cards) < 5:
            return None
        num_sequential = 1
        for i in reversed(range(len(cards) - 1)):
            if cards[i].val + 1 == cards[i + 1].val:
                num_sequential += 1
            elif cards[i].val != cards[i + 1].val:
                num_sequential = 1
            if num_sequential == 5:
                return cards[i:i + 5]
        if num_sequential == 4 and cards[0].val == 2 and cards[-1].val == 14:
            return [card.Card(cards[-1].suit, 14)] + cards[0:4]
        return None
    elif hand == HandType.THREE_KIND:
        return find_n_kind(cards, 3)
    elif hand == HandType.TWO_PAIR:
        pair_1 = find_hand(cards, HandType.PAIR)
        if pair_1 is not None:
            pair_2 = find_hand([c for c in cards if c not in pair_1],
                               HandType.PAIR)
            if pair_2 is not None:
                return pair_1 + pair_2
        return None
    elif hand == HandType.PAIR:
        return find_n_kind(cards, 2)
    elif hand == HandType.HIGH_CARD:
        return [max(cards)]


# Returns a score object if the hand can be formed, none otherwise
def try_hand(cards, hand):
    h = find_hand(cards, hand)
    if h is not None:
        return Score(hand, h, cards)


class EvalNode:
    def __init__(self, hand):
        self.hand = hand
        self.children = []
        self.best_branch_hand = None


eval_nodes = {h: EvalNode(h) for h in HandType}
eval_root = eval_nodes[HandType.HIGH_CARD]


def _link(ph, ch):
    eval_nodes[ph].children.append(eval_nodes[ch])


def best_branch_hand(eval_node):
    return max(
        [eval_node.hand] + [best_branch_hand(c) for c in eval_node.children])


_link(HandType.HIGH_CARD, HandType.FLUSH)
_link(HandType.FLUSH, HandType.STRAIGHT_FLUSH)
_link(HandType.STRAIGHT_FLUSH, HandType.ROYAL_FLUSH)
_link(HandType.HIGH_CARD, HandType.PAIR)
_link(HandType.PAIR, HandType.THREE_KIND)
_link(HandType.THREE_KIND, HandType.FOUR_KIND)
_link(HandType.THREE_KIND, HandType.FULL_HOUSE)
_link(HandType.PAIR, HandType.TWO_PAIR)
_link(HandType.HIGH_CARD, HandType.STRAIGHT)

for h in HandType:
    eval_nodes[h].best_branch_hand = best_branch_hand(eval_nodes[h])


def score(cards):
    cards.sort()
    assert len(cards) >= CARDS_TO_USE, "Must have {} cards to score".format(
        CARDS_TO_USE)

    to_expand = [eval_root]
    best_score = None
    while to_expand:
        cur = to_expand.pop()
        if best_score is None or cur.best_branch_hand > best_score.hand_type:
            to_expand.extend(cur.children)
            hand_cards = find_hand(cards, cur.hand)
            if hand_cards is not None and (best_score is None
                                           or cur.hand > best_score.hand_type):
                best_score = Score(cur.hand, hand_cards, cards)

    if best_score is not None:
        return best_score
    raise "No valid hand found"

File: scorer/scorer.go
package scorer

import (
	"fmt"
	"sort"

	"github.com/jbrunsting/poker-player/card"
)

const cardsToUse = 5

type HandType int

const (
	HandRoyalFlush    = 10
	HandStraightFlush = 9
	HandFourKind      = 8
	HandFullHouse     = 7
	HandFlush         = 6
	HandStraight      = 5
	HandThreeKind     = 4
	HandTwoPair       = 3
	HandPair          = 2
	HandHighCard      = 1
)

const (
	maxHand = 10
	minHand = 1
)

type tieResult int

const (
	lt = iota
	gt
	tie
)

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// Cards must be sorted in descending order
func tiebreaker(cardsA []card.Card, aAceHi bool, cardsB []card.Card, bAceHi bool) tieResult {
	cards := [][]card.Card{cardsA, cardsB}
	ace_hi := []bool{aAceHi, bAceHi}

	rankings := make([][]int, 2)
	for i := range rankings {
		for _, c := range cards[i] {
			v := c.Val
			if v == card.AceVal && !ace_hi[i] {
				v = card.AceLowVal
			}
			rankings[i] = append(rankings[i], v)
		}
		sort.Sort(sort.Reverse(sort.IntSlice(rankings[i])))
	}

	for i := 0; i < min(len(rankings[0]), len(rankings[1])); i++ {
		if rankings[0][i] < rankings[1][i] {
			return lt
		} else if rankings[0][i] > rankings[1][i] {
			return gt
		}
	}

	if len(rankings[0]) < len(rankings[1]) {
		return lt
	} else if len(rankings[0]) > len(rankings[1]) {
		return gt

	}

	return tie
}

type Score struct {
	Type      HandType
	HandCards []card.Card
	SideCards []card.Card
}

func handTypeToString(t HandType) string {
	if t == HandRoyalFlush {
		return "royalFlush"
	} else if t == HandStraightFlush {
		return "straightFlush"
	} else if t == HandFourKind {
		return "fourKind"
	} else if t == HandFullHouse {
		return "fullHouse"
	} else if t == HandFlush {
		return "flush"
	} else if t == HandStraight {
		return "straight"
	} else if t == HandThreeKind {
		return "threeKind"
	} else if t == HandTwoPair {
		return "twoPair"
	} else if t == HandPair {
		return "pair"
	} else if t == HandHighCard {
		return "highCard"
	}
	return "unknown"
}

func (s Score) String() string {
	typeName := ""
	if s.Type == HandRoyalFlush {
		typeName = "royalFlush"
	} else if s.Type == HandStraightFlush {
		typeName = "straightFlush"
	} else if s.Type == HandFourKind {
		typeName = "fourKind"
	} else if s.Type == HandFullHouse {
		typeName = "fullHouse"
	} else if s.Type == HandFlush {
		typeName = "flush"
	} else if s.Type == HandStraight {
		typeName = "straight"
	} else if s.Type == HandThreeKind {
		typeName = "threeKind"
	} else if s.Type == HandTwoPair {
		typeName = "twoPair"
	} else if s.Type == HandPair {
		typeName = "pair"
	} else if s.Type == HandHighCard {
		typeName = "highCard"
	}
	return typeName
}

func (s *Score) Equals(o *Score) bool {
	return !s.LessThan(o) && !o.LessThan(s)
}

func (s *Score) LessThan(o *Score) bool {
	if s.Type != o.Type {
		return s.Type < o.Type

	}
	// Tiebreaker
	sAceHi := true
	oAceHi := true
	// Ace is low iff we have a straight with the ace at the bottom (no king
	// in the straight)
	if s.Type == HandStraight && card.HasCard(s.HandCards, card.AceVal-1) {
		sAceHi = false
	}
	if o.Type == HandStraight && card.HasCard(o.HandCards, card.AceVal-1) {
		oAceHi = false

	}
	handTiebreaker := tiebreaker(s.HandCards, sAceHi, o.HandCards, oAceHi)
	if handTiebreaker != tie {
		return handTiebreaker == lt
	}

	sideTiebreaker := tiebreaker(
		s.SideCards,
		false,
		o.SideCards,
		false,
	)
	return sideTiebreaker == lt
}

func findNKind(cards []card.Card, n int) []card.Card {
	numSequential := 1
	for i := len(cards) - 2; i >= 0; i-- {
		if cards[i].Val == cards[i+1].Val {
			numSequential++
		} else {
			numSequential = 1
		}
		if numSequential == n {
			return cards[i : i+n]
		}
	}
	return []card.Card{}
}

func filterCards(cards []card.Card, filter func(card.Card) bool) []card.Card {
	filtered := []card.Card{}
	for _, c := range cards {
		if filter(c) {
			filtered = append(filtered, c)
		}
	}
	return filtered
}

func maxVal(cards []card.Card) int {
	maxVal := cards[0].Val
	for _, c := range cards {
		if c.Val > maxVal {
			maxVal = c.Val
		}
	}
	return maxVal
}

// Returns empty list if not found
func findHand(cards []card.Card, hand HandType) []card.Card {
	sort.Slice(cards, func(i, j int) bool {
		return cards[i].Val > cards[j].Val
	})
	return _findHand(cards, hand)
}

func _findHand(cards []card.Card, hand HandType) []card.Card {
	if hand == HandRoyalFlush {
		royalCards := filterCards(cards, func(c card.Card) bool {
			return c.Val >= 10
		})
		return findHand(royalCards, HandStraightFlush)
	} else if hand == HandStraightFlush {
		// We do this whole bestStraight thing to account for cases where we
		// have 10 cards or more, since in those cases we could have multiple
		// suits with a straight flush and we should pick the best one
		bestStraight := []card.Card{}
		for suit := 0; suit < card.NumSuits; suit++ {
			suitCards := filterCards(cards, func(c card.Card) bool {
				return c.Suit == suit
			})
			straightCards := findHand(suitCards, HandStraight)
			if len(straightCards) != 0 &&
				(len(bestStraight) == 0 || maxVal(bestStraight) < maxVal(straightCards)) {
				bestStraight = straightCards
			}
		}
		return bestStraight
	} else if hand == HandFourKind {
		return findNKind(cards, 4)
	} else if hand == HandFullHouse {
		threeKindCards := findHand(cards, HandThreeKind)
		if len(threeKindCards) != 0 {
			otherCards := filterCards(cards, func(c card.Card) bool {
				for _, o := range threeKindCards {
					if c == o {
						return false
					}
				}
				return true
			})
			pairCards := findHand(otherCards, HandPair)
			if len(pairCards) != 0 {
				return append(threeKindCards, pairCards...)
			}
		}
		return []card.Card{}
	} else if hand == HandFlush {
		bestFlush := []card.Card{}
		for suit := 0; suit < card.NumSuits; suit++ {
			suitCards := filterCards(cards, func(c card.Card) bool {
				return c.Suit == suit
			})
			if len(suitCards) >= 5 {
				flushCards := suitCards[:5]
				if len(flushCards) != 0 && (len(bestFlush) == 0 || maxVal(bestFlush) < maxVal(flushCards)) {
					bestFlush = flushCards
				}
			}
		}
		return bestFlush
	} else if hand == HandStraight {
		if len(cards) < 5 {
			return []card.Card{}
		}
		numSequential := 1
		for i := 0; i < len(cards)-1; i++ {
			if cards[i].Val == cards[i+1].Val+1 {
				numSequential++
			} else if cards[i].Val != cards[i+1].Val {
				numSequential = 1
			}
			if numSequential == 5 {
				return cards[i-3 : i+2]
			}
		}
		if numSequential == 4 && cards[0].Val == card.MinVal && cards[len(cards)-1].Val == card.AceVal {
			return append([]card.Card{card.Card{cards[len(cards)-1].Suit, card.AceVal}}, cards[0:4]...)
		}
		return nil
	} else if hand == HandThreeKind {
		return findNKind(cards, 3)
	} else if hand == HandTwoPair {
		pair1Cards := findHand(cards, HandPair)
		if len(pair1Cards) != 0 {
			otherCards := filterCards(cards, func(c card.Card) bool {
				for _, o := range pair1Cards {
					if c == o {
						return false
					}
				}
				return true
			})
			pair2Cards := findHand(otherCards, HandPair)
			if len(pair2Cards) != 0 {
				return append(pair1Cards, pair2Cards...)
			}
		}
		return []card.Card{}
	} else if hand == HandPair {
		return findNKind(cards, 2)
	} else if hand == HandHighCard {
		return cards[0:1]
	}

	return []card.Card{}
}

func GetScore(cards []card.Card) Score {
	for i1, o1 := range cards {
		for i2, o2 := range cards {
			if i1 != i2 && o1 == o2 {
				panic(fmt.Sprintf("DUPLICATE for cards %v, card %s at %d, %d", cards, o1, i1, i2))
			}
		}
	}
	for i := maxHand + 1; i >= minHand; i-- {
		handType := HandType(i)
		handCards := findHand(cards, handType)
		sideCards := filterCards(cards, func(c card.Card) bool {
			for _, o := range handCards {
				if c == o {
					return false
				}
			}
			return true
		})
		if len(handCards) != 0 {
			return Score{handType, handCards, sideCards[:cardsToUse-len(handCards)]}
		}
	}
	panic("No score found")
}
